### PR TITLE
Update baseline and recompiled cut_in_mid_rear

### PR DESCRIPTION
**Pull request by xuhan417**
# PR Details
## Description
This is UCLA team's commit of the IHP2 platoon plugin cut-in join from middle and rear.

## Related Issue
There is no related issues.

## Motivation and Context
This is the development for IHP2 project cut-in join from middle and rear.

PR context: 
- Mid/rear join functions.
- Include "join index" in STATUS params, and keep tracking it as a property of platoonmember class.
- Update plan type name and platoon manager APF function.

## How Has This Been Tested?
- Add unit test to test additional state transitions.
- Compile and ran on UCLA's vehicle platform (single vehicle only).

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: platooning_strategic_IHP/src/platoon_strategic_ihp.cpp [hunk 1]
```diff
@@ -2234,6 +2360,9 @@ namespace platoon_strategic_ihp
             return;
         }
         
+        // UCLA: read current request plan
```
Comment (TonysCousin) [inline platooning_strategic_IHP/src/platoon_strategic_ihp.cpp:2363]: Why is this here?  It isn't used anywhere.

File: platooning_strategic_IHP/src/platoon_strategic_ihp.cpp [hunk 2]
```diff
@@ -2528,7 +2664,7 @@ namespace platoon_strategic_ihp
             fmter %pose_ecef_point_.ecef_x;     //  index = 2
             fmter %pose_ecef_point_.ecef_y;     //  index = 3
             fmter %pose_ecef_point_.ecef_z;     //  index = 4   
-            fmter %join_index;                  //  index = 5
+            //Note: remove join_index to STATUS param. request.join_index = -1;
```
Comment (TonysCousin) [inline platooning_strategic_IHP/src/platoon_strategic_ihp.cpp:2667]: Need to keep the previous code hre to add join_index to the status params.

File: platooning_strategic_IHP/src/platoon_strategic_ihp.cpp [hunk 3]
```diff
@@ -2703,16 +2840,19 @@ namespace platoon_strategic_ihp
         fmter %pose_ecef_point_.ecef_x;     //  index = 2
         fmter %pose_ecef_point_.ecef_y;     //  index = 3
         fmter %pose_ecef_point_.ecef_z;     //  index = 4
-             
-        fmter %target_join_index_;             //  index = 5
+        //Note: remove join_index to STATUS param. request.join_index = target_join_index_;
```
Comment (TonysCousin) [inline platooning_strategic_IHP/src/platoon_strategic_ihp.cpp:2843]: Need the join_index as part of the params

File: platooning_strategic_IHP/src/platoon_strategic_ihp.cpp [hunk 4]
```diff
@@ -2692,6 +2828,7 @@ namespace platoon_strategic_ihp
         // UCLA: assign a new plan type
```
Comment (TonysCousin) [inline platooning_strategic_IHP/src/platoon_strategic_ihp.cpp:2828]: 10 lines above here target_join_index is calculated by asking PlatoonManager about the index of platoon members. Problem is pm_ represents the hosts current platoon (as a solo vehicle), and not the platoon it desires to join.  So this info is meaningless.
I think what we need is for platoon_strategic_ihp to be able to look at INFO message from the neighbor platoon leader and from its leader DTD, size and length figure out the closest index.  As I have suspected for some time, the INFO message probably really needs to include something about each member (e.g. list of IDs in order); then a potential joiner can monitor the STATUS messages coming from each member to get their exact locations.  That would be a major change, so is worth discussing; we can probably come up with a better approach after some careful thought.

File: platooning_strategic_IHP/src/platoon_strategic_ihp.cpp [hunk 5]
```diff
@@ -2703,16 +2840,19 @@ namespace platoon_strategic_ihp
...
 
         request.strategy_params = fmter.str();
-                
+        
         mobility_request_publisher_(request); 
         ROS_DEBUG_STREAM("Published Mobility cut-in join request to the leader");
         ROS_WARN("Published Mobility cut-in join request to the leader");
     }
 
+    // UCLA: run prepare to depart state for depart from platoon
```
Comment (TonysCousin) [inline platooning_strategic_IHP/src/platoon_strategic_ihp.cpp:2852]: I suspect that there is no transition out of the PREPARETOJOIN state other than for exception situations.  I would expect to see it in the run_prepare_to_join() method and normally transition to FOLLOWER or CANDIDATEFOLLOWER.  Maybe I'm looking in the wrong place?

**Review by TonysCousin — CHANGES_REQUESTED** [review]

File: platooning_strategic_IHP/include/platoon_manager_ihp.h [hunk 6]
```diff
@@ -162,7 +165,7 @@ namespace platoon_strategic_ihp
          * \param platoonId sender platoon id
          * \param params strategy params from STATUS message in the format of "CMDSPEED:xx,DOWNTRACK:xx,SPEED:xx"
          **/
-        void updatesOrAddMemberInfo(std::string senderId, double cmdSpeed, double dtDistance, double ctDistance, double curSpeed);
+        void updatesOrAddMemberInfo(std::string senderId, double cmdSpeed, double dtDistance, double ctDistance, double curSpeed, int joinIndex);
```
[the remark this answers is not in the document]
Comment (xuhan417) [reply platooning_strategic_IHP/include/platoon_manager_ihp.h:168]: Switch join index to STATUS param.

File: platooning_strategic_IHP/src/platoon_strategic_ihp.cpp [hunk 7]
```diff
@@ -2149,9 +2271,7 @@ namespace platoon_strategic_ihp
         cav_msgs::PlanType plan_type = msg.plan_type;
         bool isCreatingGap = plan_type.type == cav_msgs::PlanType::PLATOON_CUTIN_JOIN;
         bool isFinishLaneChangeFront = plan_type.type == cav_msgs::PlanType::CUT_IN_FRONT_DONE; 
-        // todo: add cut-in from middle and rear 
-        // bool isFinishLaneChangeMid
-        // bool isFinishLaneChangeRear
+        bool isFinishLaneChangeMidorRear = plan_type.type == cav_msgs::PlanType::CUT_IN_MID_OR_REAR_DONE;
```
[the remark this answers is not in the document]
Comment (xuhan417) [reply platooning_strategic_IHP/src/platoon_strategic_ihp.cpp:2302]: updated

**Review by saina-ramyar — CHANGES_REQUESTED** [review]
Some of my comments in this PR are from before, and they are already addressed or disusssed. You can ignore those. 
Also, I noticed these files: src/.vscde and src/reference_code.zip which seem to be auto-generated by vscde. These shouldn't be merged, so please delete them, and make sure such files don't get pushed to the PR.

**Review by TonysCousin — APPROVED** [review]

**Review by saina-ramyar — APPROVED** [review]

**Comment by saina-ramyar** [comment]
Replaced by PR #1719 due to conflicts.

**Pull request closed without merging**